### PR TITLE
Refactor instruction buffer

### PR DESCRIPTION
**Pull request by michaelhly**

### CODE REVIEW

File: programs/smart-wallet/src/instructions/buffer_finalize.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
+//! Init or close an [InstructionBuffer].
```
Comment (macalinao) [inline programs/smart-wallet/src/instructions/buffer_finalize.rs:1]: This doc is confusing.

File: programs/smart-wallet/src/instructions/buffer_finalize.rs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+#[derive(Accounts)]
+pub struct FinalizeBuffer<'info> {
+    #[account(mut)]
+    pub buffer: Account<'info, InstructionBuffer>,
+    pub smart_wallet: Account<'info, SmartWallet>,
+    pub transaction: Box<Account<'info, Transaction>>,
+    pub owner: Signer<'info>,
+}
+
+/// Emitted when a [InstructionBuffer] is initialized.
```
Comment (macalinao) [inline programs/smart-wallet/src/instructions/buffer_finalize.rs:14]: This doc is confusing.

File: programs/smart-wallet/src/instructions/buffer_finalize.rs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    Ok(())
+}
+
+impl<'info> Validate<'info> for FinalizeBuffer<'info> {
+    fn validate(&self) -> Result<()> {
+        invariant!(self.buffer.finalized_at == 0);
+
+        invariant!(
+            self.smart_wallet.owner_set_seqno == self.transaction.owner_set_seqno,
```
Comment (macalinao) [inline programs/smart-wallet/src/instructions/buffer_finalize.rs:41]: Move this to execution

File: programs/smart-wallet/src/instructions/buffer_execute.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
+//! Executes an instruction off of the [InstructionBuffer].
+
+use anchor_lang::solana_program::program::invoke_signed;
+
+use crate::*;
+
+#[derive(Accounts)]
+pub struct ExecuteIx<'info> {
```
Comment (macalinao) [inline programs/smart-wallet/src/instructions/buffer_execute.rs:8]: Rename the mod to `buffer_execute_ix` and the struct to `BufferExecuteIX`

File: tests/loader.spec.ts [hunk 5]
```diff
@@ -118,11 +154,33 @@ describe("instruction loader", () => {
     expect(bufferData.execCount).to.be.eq(3);
   });
 
-  it("Close instruction buffer", async () => {
-    const tx = sdk.instructionLoader.closeBuffer(bufferAccount);
-    await expectTXTable(tx, "close buffer").to.be.fulfilled;
+  it("Cannot execute on unfalized buffer", async () => {
```
Comment (macalinao) [inline tests/loader.spec.ts:157]: typo `unfalized`

**Review by macalinao — CHANGES_REQUESTED** [review]

File: programs/smart-wallet/src/state.rs [hunk 6]
```diff
@@ -200,8 +200,23 @@ pub struct SubaccountInfo {
...
-    /// Key that can write to the buffer.
+    /// Sequence of the ownership set.
+    ///
+    /// This may be used to see if the owners on the multisig have changed
+    pub owner_set_seqno: u32,
+    /// - If set to [crate::NO_ETA], the instructions in the [InstructionBuffer::instructions] may be executed at any time.
+    /// - Otherwise, instructions may be executed at any point after the ETA has elapsed.
+    pub eta: i64,
+    /// Time denoting when buffer is ready to be executed.
+    pub finalized_at: i64,
```
Comment (macalinao) [inline programs/smart-wallet/src/state.rs:211]: Delete this, and replace the `finalized` check with checking if `writer` is `system_program::ID`.

File: programs/smart-wallet/src/state.rs [hunk 7]
```diff
@@ -200,8 +200,23 @@ pub struct SubaccountInfo {
...
+    /// Role that can set the buffer writer and executer.
+    pub admin: Pubkey,
```
Comment (macalinao) [inline programs/smart-wallet/src/state.rs:213]: Delete this `admin`.

Instead, the `writer` should be able to modify the `executor`.

File: programs/smart-wallet/src/instructions/buffer_execute_ix.rs [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        b"GokiSmartWallet" as &[u8],
+        &smart_wallet.base.to_bytes(),
+        &[smart_wallet.bump],
+    ]];
+
+    let buffer = &mut ctx.accounts.buffer;
+    let ix = &buffer.instructions[buffer.exec_count as usize];
+    invoke_signed(&ix.into(), ctx.remaining_accounts, wallet_seeds)?;
+
+    buffer.exec_count += 1;
```
Comment (macalinao) [inline programs/smart-wallet/src/instructions/buffer_execute_ix.rs:29]: change this to `checked_add`

File: programs/smart-wallet/src/instructions/buffer_execute_ix.rs [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+    Ok(())
+}
+
+impl<'info> Validate<'info> for ExecuteBufferIX<'info> {
+    fn validate(&self) -> Result<()> {
+        assert_keys_eq!(self.buffer.executer, self.executor.key());
```
Comment (macalinao) [inline programs/smart-wallet/src/instructions/buffer_execute_ix.rs:36]: `executer` typo

File: programs/smart-wallet/src/instructions/buffer_execute_ix.rs [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+pub struct ExecuteBufferIX<'info> {
+    #[account(mut)]
+    pub buffer: Account<'info, InstructionBuffer>,
+    /// The buffer's [SmartWallet]
+    pub smart_wallet: Account<'info, SmartWallet>,
+    /// The [InstructionBuffer::executer].
+    pub executor: Signer<'info>,
+}
+
+pub fn handler<'info>(ctx: Context<'_, '_, '_, 'info, ExecuteBufferIX<'info>>) -> Result<()> {
```
Comment (macalinao) [inline programs/smart-wallet/src/instructions/buffer_execute_ix.rs:17]: You should allow for an instruction `index` in the arguments so the transactions can be parallelized.

File: programs/smart-wallet/src/instructions/buffer_finalize.rs [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    /// The buffer.
+    pub buffer: Pubkey,
+    /// The time [InstructionBuffer] is finalized.
+    #[index]
+    pub time: i64,
+}
+
+pub fn handler(ctx: Context<FinalizeBuffer>) -> Result<()> {
+    let buffer = &mut ctx.accounts.buffer;
+    buffer.finalized_at = Clock::get()?.unix_timestamp;
```
Comment (macalinao) [inline programs/smart-wallet/src/instructions/buffer_finalize.rs:24]: Delete `finalized_at` and replace with changing the `writer` to `system_program::ID`.

File: programs/smart-wallet/src/instructions/buffer_close.rs [hunk 12]
```diff
@@ -0,0 +1,36 @@
+//! Closes an [InstructionBuffer] for execution.
+
+use crate::*;
+
+#[derive(Accounts)]
+pub struct CloseBuffer<'info> {
+    #[account(mut, close = executer)]
+    pub buffer: Account<'info, InstructionBuffer>,
+    pub executer: Signer<'info>,
```
Comment (macalinao) [inline programs/smart-wallet/src/instructions/buffer_close.rs:9]: rename to `writer_or_executor`

File: programs/smart-wallet/src/instructions/buffer_execute_ix.rs [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    Ok(())
+}
+
+impl<'info> Validate<'info> for ExecuteBufferIX<'info> {
+    fn validate(&self) -> Result<()> {
+        assert_keys_eq!(self.buffer.executer, self.executor.key());
+        assert_keys_eq!(self.buffer.smart_wallet, self.smart_wallet.key());
+
+        invariant!(self.buffer.finalized_at > 0, BufferNotFinalized);
+        invariant!(usize::from(self.buffer.exec_count) < self.buffer.instructions.len());
```
Comment (macalinao) [inline programs/smart-wallet/src/instructions/buffer_execute_ix.rs:40]: Move this into a method on `InstructionBuffer`, called `has_fully_executed`.

Also, this should throw an `ErrorCode` for better UX.

File: programs/smart-wallet/src/instructions/buffer_set_role.rs [hunk 14]
```diff
@@ -0,0 +1,90 @@
+use vipers::program_err;
```
Comment (macalinao) [inline programs/smart-wallet/src/instructions/buffer_set_role.rs:1]: delete file

File: programs/smart-wallet/src/instructions/buffer_write.rs [hunk 15]
```diff
@@ -3,39 +3,36 @@
 use crate::*;
```
Comment (macalinao) [inline programs/smart-wallet/src/instructions/buffer_write.rs:3]: Rename to `buffer_write_ix`.

File: programs/smart-wallet/src/state.rs [hunk 16]
```diff
@@ -200,8 +200,23 @@ pub struct SubaccountInfo {
...
+    /// Role that can write to the buffer.
     pub writer: Pubkey,
-    /// Staged instructions to be executed.
-    pub staged_tx_instructions: Vec<TXInstruction>,
+    /// Role that can execute instructions off the buffer.
+    pub executer: Pubkey,
+    /// Smart wallet the buffer belongs to.
+    pub smart_wallet: Pubkey,
+    /// Vector of instructions.
+    pub instructions: Vec<TXInstruction>,
```
Comment (macalinao) [inline programs/smart-wallet/src/state.rs:221]:
```
pub txs: Vec<TXInstructionBundle>
```

and

```
pub struct TXInstructionBundle {
  pub instructions: Vec<TXInstruction>,
  pub is_executed: bool
}
```

**Review by macalinao — CHANGES_REQUESTED** [review]

File: programs/smart-wallet/src/state.rs [hunk 17]
```diff
@@ -198,10 +199,67 @@ pub struct SubaccountInfo {
...
+
+        Ok(())
+    }
+}
+
+/// Container holding a bundle of instructions.
+#[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug, Default, PartialEq)]
+pub struct InstructionBundle {
+    /// Execution counter on the [InstructionBundle].
     pub exec_count: u8,
```
Comment (macalinao) [inline programs/smart-wallet/src/state.rs:202]: Why would a bundle be executed more than once?

File: programs/smart-wallet/src/instructions/buffer_execute_ix.rs [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+        b"GokiSmartWallet" as &[u8],
+        &smart_wallet.base.to_bytes(),
+        &[smart_wallet.bump],
+    ]];
+
+    let buffer = &mut ctx.accounts.buffer;
+    let mut bundle = unwrap_opt!(buffer.get_bundle(bundle_index), BufferBundleNotFound);
+    invariant!(!bundle.is_executed(), BufferBundleExecuted);
+
+    let ix = &bundle.instructions[usize::from(bundle.exec_count)];
```
Comment (macalinao) [inline programs/smart-wallet/src/instructions/buffer_execute_ix.rs:33]: Let's have bundles be executed atomically as one transaction. No exec_count

**Review by macalinao — CHANGES_REQUESTED** [review]
Bundles should only be executed atomically, to keep things simple

File: programs/smart-wallet/src/state.rs (hunk 17 above)
Comment (michaelhly) [reply programs/smart-wallet/src/state.rs:202]: This is the counter to interate through the instructions. I can rename if you have suggestions.

File: programs/smart-wallet/src/instructions/buffer_execute_ix.rs (hunk 18 above)
Comment (michaelhly) [reply programs/smart-wallet/src/instructions/buffer_execute_ix.rs:33]: We should keep the instruction small.

**Pull request merged**